File: api/app/Services/Mcp/TemplateEngine.php
<?php

namespace App\Services\Mcp;

use Symfony\Component\Yaml\Yaml;
use Twig\Environment;
use Twig\Loader\FilesystemLoader;

/**
 * Port of app/template_engine.py — loads template bundles, renders *.j2 files
 * with user variables, copies other files to the build context.
 */
class TemplateEngine
{
    public function __construct(
        private readonly string $templatesDir,
        private readonly string $serversDir,
    ) {}

    /** @return array<int, array<string, mixed>> */
    public function listTemplates(): array
    {
        if (! is_dir($this->templatesDir)) {
            return [];
        }
        $out = [];
        $entries = scandir($this->templatesDir) ?: [];
        sort($entries);
        foreach ($entries as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $metaPath = $this->templatesDir."/{$entry}/template.yaml";
            if (is_file($metaPath)) {
                $meta = $this->loadMeta($metaPath);
                if ($meta) {
                    $out[] = $meta;
                }
            }
        }
        return $out;
    }

    public function getTemplate(string $name): ?array
    {
        $metaPath = $this->templatesDir."/{$name}/template.yaml";
        if (! is_file($metaPath)) {
            return null;
        }
        return $this->loadMeta($metaPath);
    }

    /**
     * Render a template bundle into the server's build-context directory.
     * Returns the absolute output directory path.
     *
     * @param array<string, string> $config
     */
    public function render(string $templateName, string $serverName, array $config): string
    {
        $templateDir = $this->templatesDir."/{$templateName}";
        if (! is_dir($templateDir)) {
            throw new \InvalidArgumentException("Template '{$templateName}' not found");
        }

        $meta = $this->loadMeta($templateDir.'/template.yaml');
        if (! $meta) {
            throw new \InvalidArgumentException("Invalid template metadata for '{$templateName}'");
        }

        $variables = ['server_name' => $serverName];
        foreach ($meta['variables'] as $var) {
            if (array_key_exists($var['name'], $config)) {
                $variables[$var['name']] = $config[$var['name']];
            } elseif ($var['default'] !== null) {
                $variables[$var['name']] = $var['default'];
            } elseif ($var['required']) {
                throw new \InvalidArgumentException("Required variable '{$var['name']}' not provided");
            }
        }

        $outputDir = $this->serversDir."/{$serverName}";
        if (! is_dir($outputDir)) {
            mkdir($outputDir, 0755, true);
        }

        $twig = new Environment(new FilesystemLoader($templateDir), [
            'autoescape' => false,
            'strict_variables' => false,
        ]);

        foreach (glob($templateDir.'/*.j2') as $j2File) {
            $name = basename($j2File);
            $rendered = $twig->render($name, $variables);
            $outName = substr($name, 0, -3); // strip .j2
            file_put_contents($outputDir.'/'.$outName, $rendered);
        }

        foreach (scandir($templateDir) as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $src = $templateDir.'/'.$entry;
            if (! is_file($src)) {
                continue;
            }
            $ext = pathinfo($entry, PATHINFO_EXTENSION);
            if ($ext === 'j2' || $ext === 'yaml') {
                continue;
            }
            copy($src, $outputDir.'/'.$entry);
        }

        return $outputDir;
    }

    public function cleanup(string $serverName): void
    {
        $dir = $this->serversDir."/{$serverName}";
        if (is_dir($dir)) {
            $this->rrmdir($dir);
        }
    }

    private function loadMeta(string $path): ?array
    {
        try {
            $data = Yaml::parseFile($path);
        } catch (\Throwable) {
            return null;
        }
        if (! is_array($data) || empty($data['name'])) {
            return null;
        }

        $vars = [];
        foreach ($data['variables'] ?? [] as $v) {
            if (! is_array($v) || empty($v['name'])) {
                continue;
            }
            $vars[] = [
                'name' => $v['name'],
                'description' => $v['description'] ?? '',
                'default' => array_key_exists('default', $v) ? (string) $v['default'] : null,
                'required' => (bool) ($v['required'] ?? false),
            ];
        }

        return [
            'name' => $data['name'],
            'description' => $data['description'] ?? '',
            'variables' => $vars,
        ];
    }

    private function rrmdir(string $dir): void
    {
        foreach (scandir($dir) as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $path = $dir.'/'.$entry;
            is_dir($path) ? $this->rrmdir($path) : @unlink($path);
        }
        @rmdir($dir);
    }
}

File: api/app/services/template_engine.py
"""Load template bundles (template.yaml + *.j2), render with user variables,
copy non-template files to the build context."""
from __future__ import annotations

import shutil
from pathlib import Path

import yaml
from jinja2 import Environment, FileSystemLoader, StrictUndefined  # noqa: F401 - StrictUndefined available for caller use


class TemplateEngine:
    def __init__(self, templates_dir: Path | str, servers_dir: Path | str):
        self.templates_dir = Path(templates_dir)
        self.servers_dir = Path(servers_dir)

    def list_templates(self) -> list[dict]:
        if not self.templates_dir.is_dir():
            return []
        out: list[dict] = []
        for entry in sorted(self.templates_dir.iterdir()):
            meta_path = entry / "template.yaml"
            if meta_path.is_file():
                meta = self._load_meta(meta_path)
                if meta:
                    out.append(meta)
        return out

    def get_template(self, name: str) -> dict | None:
        meta_path = self.templates_dir / name / "template.yaml"
        if not meta_path.is_file():
            return None
        return self._load_meta(meta_path)

    def render(self, template_name: str, server_name: str, config: dict[str, str]) -> Path:
        template_dir = self.templates_dir / template_name
        if not template_dir.is_dir():
            raise ValueError(f"Template '{template_name}' not found")
        meta = self._load_meta(template_dir / "template.yaml")
        if not meta:
            raise ValueError(f"Invalid template metadata for '{template_name}'")

        variables: dict[str, str] = {"server_name": server_name}
        for var in meta["variables"]:
            name = var["name"]
            if name in config:
                variables[name] = config[name]
            elif var["default"] is not None:
                variables[name] = var["default"]
            elif var["required"]:
                raise ValueError(f"Required variable '{name}' not provided")

        output_dir = self.servers_dir / server_name
        output_dir.mkdir(parents=True, exist_ok=True)

        env = Environment(
            loader=FileSystemLoader(str(template_dir)),
            autoescape=False,
            keep_trailing_newline=True,
        )

        for j2 in template_dir.glob("*.j2"):
            rendered = env.get_template(j2.name).render(**variables)
            out_name = j2.name[:-3]  # strip .j2
            (output_dir / out_name).write_text(rendered, encoding="utf-8")

        for entry in template_dir.iterdir():
            if not entry.is_file():
                continue
            ext = entry.suffix.lower().lstrip(".")
            if ext in ("j2", "yaml"):
                continue
            shutil.copy2(entry, output_dir / entry.name)

        return output_dir

    def cleanup(self, server_name: str) -> None:
        d = self.servers_dir / server_name
        if d.is_dir():
            shutil.rmtree(d, ignore_errors=True)

    def _load_meta(self, path: Path) -> dict | None:
        try:
            data = yaml.safe_load(path.read_text(encoding="utf-8"))
        except (OSError, yaml.YAMLError):
            return None
        if not isinstance(data, dict) or not data.get("name"):
            return None
        variables: list[dict] = []
        for v in data.get("variables", []) or []:
            if not isinstance(v, dict) or not v.get("name"):
                continue
            variables.append({
                "name": v["name"],
                "description": v.get("description", ""),
                "default": (None if "default" not in v else str(v["default"])),
                "required": bool(v.get("required", False)),
            })
        return {
            "name": data["name"],
            "description": data.get("description", ""),
            "variables": variables,
        }
